File: go/linkedlist/92_reverse_between.go
package linkedlist

// reverseBetween 反转链表 II
func reverseBetween(head *ListNode, left int, right int) *ListNode {

	if head == nil {
		return nil
	}

	// 不用	反转
	if left == right {
		return head
	}

	// 穿针引线法
	//  left = 2, right = 4
	// A -> B -> C -> D -> E
	// 1. 先找到left 的前一个节点 prev 永远保持不变
	// 2. 当i = left 时， cur = B, next = C
	// 3. 把 c 插入到B 的前面 (即在prev 后面)
	// 4. 把 d 插入到c 的前面

	// 哨兵节点
	dummy := &ListNode{}
	dummy.Next = head

	prev := dummy

	// after loop finished, prev 在left 前面
	// 不能用i < left -2; 因为可能 left == 1
	for i := 0; i < left-1; i++ {
		prev = prev.Next
	}

	cur := prev.Next

	// 遍历次数到right 前一位
	// 几次遍历之后，其实cur 永远指向B
	for i := 0; i < right-left; i++ {
		next := cur.Next

		// 穿针引线
		cur.Next = next.Next  // B -> D
		next.Next = prev.Next // 永远插入到prev 的下一位，C -> B
		prev.Next = next      // A -> C
	}

	return dummy.Next
}

File: rust/src/linklist/_92_reverse_between.rs
use super::ListNode;

/// 反转链表 II
pub fn reverse_between(
    head: Option<Box<ListNode>>,
    left: i32,
    right: i32,
) -> Option<Box<ListNode>> {
    if head.is_none() {
        return None;
    }

    if left == right {
        return head;
    }
    // left 2, right  4
    // A -> B -> C -> D -> E
    // 1. A -> C -> B -> D - > E
    // 2. A -> D -> C -> B - > E

    // 穿针引线法  一次遍历
    let mut dummy = Some(Box::new(ListNode::new(0)));
    dummy.as_mut().unwrap().next = head;

    let mut prev = &mut dummy;

    for _ in 0..left - 1 {
        prev = &mut prev.as_mut().unwrap().next;
    }

    let mut cur = prev.as_mut().unwrap().next.take();

    for _ in 0..right - left {
        let mut next = &mut cur.as_mut().unwrap().next.take();

        cur.as_mut().unwrap().next = next.as_mut().unwrap().next.take();
        // 此处将next 插入到prev 后面，但是prev 在 上面cur 处被切断，导致链表 以cur 为中心分成了两个部分
        next.as_mut().unwrap().next = prev.as_mut().unwrap().next.take();
        prev.as_mut().unwrap().next = next.take();
    }
    // 上述遍历完形成两个链表  A->D->C + B->E
    // 将prev 移到C
    while prev.as_mut().unwrap().next.is_some() {
        prev = &mut prev.as_mut()?.next;
    }
    // 重新连接上
    prev.as_mut().unwrap().next = cur.take();

    dummy.unwrap().next
}
